### PR TITLE
fix(ledger): refund the pool deposit each registration holds

### LINKED ISSUE #3670: ledger: refund the pool deposit held by each registration
## Problem

Pool retirement refunds the current protocol parameter's pool deposit instead of the amount retained for the effective registration being retired. A parameter change between registration and retirement can therefore create or destroy ledger value.

## Acceptance criteria

- Persist the deposit amount held by each effective pool registration.
- Define repeated-registration semantics without replacing the refundable amount incorrectly.
- Refund exactly the retained amount when the registration retires.
- Provide deterministic migration and backfill behavior for legacy registrations.
- Preserve the held amount through rollback, replay, and restart.
- Add increasing, decreasing, repeated-registration, migration, rollback, and replay regressions through the production storage path.
- Pass focused/full normal and race tests, metadata backend checks, conformance, lint, vet, and documentation parity.

### PR DESCRIPTION
**Pull request by wolf31o2**
Track retained pool deposits across re-registrations and migration replay.

- Preserve held deposits across re-registration; reset after completed reap.
- Reconstruct legacy values from registration, retirement, and epoch history.
- Validate replayed ADD COLUMN types.
- Update database documentation.

Tests: `go test ./database/plugin/metadata/sqlstore/migrations/... -count=1`

Fixes #3670

### CODE REVIEW

File: database/plugin/metadata/sqlstore/migrations/v7/sqlite/expand.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+-- The pool deposit a registration retains, as distinct from `deposit_amount`,
+-- which is what the block era's certificate deposit function computed from the
+-- protocol parameters in force at that registration's slot. The two differ for
+-- a re-registration: cardano-ledger's POOL rule charges a deposit only when the
+-- pool is not already registered, so a re-registration leaves `psDeposits`
+-- alone and the pool keeps holding the deposit its first registration paid.
+-- POOLREAP refunds `psDeposits`, so refunding `deposit_amount` created or
+-- destroyed ledger value whenever a poolDeposit parameter change landed between
+-- a pool's first and last registration.
+ALTER TABLE `pool_registration` ADD COLUMN `deposit_held` text;
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply database/plugin/metadata/sqlstore/migrations/v7/sqlite/expand.sql]: Valid. The expand phase is not one transaction: execDDL runs each statement on its own connection in autocommit, and the phase advance is a separate write afterwards, so a stop in between replays the whole phase. Versions 2, 5 and 7 all add columns and all failed that replay with a duplicate-column error before their data statements ran.

Fixed in the shared runner rather than in v7's SQL. A migration resource cannot carry the guard itself: resources are authored once in SQLite syntax and translated per dialect, and SQLite has no ADD COLUMN IF NOT EXISTS. execDDL now inspects the live schema when an ALTER TABLE ADD COLUMN fails (pragma_table_info on SQLite, information_schema.columns on PostgreSQL and MySQL) and skips the statement only when that column is present. Keying on the schema rather than driver error text means no dialect is matched by message and an ALTER that failed for any other reason still fails the migration. No migration SQL changed, so no checksum changed.

The blast radius is all seven versions, so the fix is covered as a class:

- TestRunnerReplaysEveryShippedVersionFromExpand rewinds each shipped version to phase=expand against a database that already has its full effect applied and re-runs. Without the fix, v2, v5 and v7 fail and v1, v3, v4 and v6 pass, which is the exact class.
- TestDepositHeldExpandPhaseReplaysAfterInterruptedUpgrade rewinds v7 specifically with the column present and deposit_held reset to NULL, then asserts the backfill ran and the version reached complete.
- TestRunnerReportsAddColumnFailureOnMissingTable and TestRunnerDoesNotSkipNonAddColumnFailure pin the non-masking direction, including a failing statement on a table that already carries the named column.
- TestAddColumnPatternMatchesShippedMigrations pins the guard against dialect-translation drift: all seven ADD COLUMN statements in each of the three registries are recognised, plus unquoted and multi-line forms, while ADD CONSTRAINT, DROP COLUMN, RENAME COLUMN, CREATE TABLE, CREATE INDEX and UPDATE are not.

DATABASE.md records the restart semantics.

Scope limits, stated in the doc comment: existence is checked, not declared type, so a hand-added same-named column of another type would be accepted; and PostgreSQL and MySQL replay is covered by statement-shape unit tests only, not against live servers.

File: database/plugin/metadata/sqlstore/pool_deposit_held_test.go [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+	require.Equal(t, "800", depositHeldColumn(t, store, pool, 1_100))
+	require.Equal(
+		t,
+		uint64(800),
+		depositHeldRefund(t, store, pool, 3, 3_000),
+	)
+}
+
+// Repeated registrations carry the first registration's amount forward through
+// every later one, including two in the same block distinguished only by
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply database/plugin/metadata/sqlstore/pool_deposit_held_test.go]: Valid. Every certificate in that test is at a distinct slot, so the block_index tie-break was not exercised.

The specific case the comment described is also unrepresentable: two registrations of one pool cannot both persist in a block, because poolRegistrationInsertSQL resolves the unique (pool_id, added_slot) key with ON CONFLICT DO NOTHING. The comment now says that instead of claiming the coverage.

The tie-break is reachable between retirement certificates, since pool_retirement has no unique (pool_id, added_slot) key, and TestPoolDepositHeldSameBlockRetirementsOrderByBlockIndex now covers it: a registration at slot 100 holding 500, two retirement certificates in the block at slot 200 naming epochs 1 and 4, then a re-registration at slot 2_500 charged 900. Slot 2_500 is in epoch 2, so the effective epoch-4 retirement is still pending and the re-registration carries 500 forward; the superseded epoch-1 retirement was reaped at slot 1_000, so reading it instead makes the re-registration look post-reap and charge 900.

Proved load-bearing in both places that resolve the tie. Removing COALESCE(t.block_index, 0) DESC from latestPoolRetirementBefore's ORDER BY makes deposit_held at slot 2_500 read 900 instead of 500. Removing it from GetPoolsRetiringAtEpoch's latest_ret window makes the pool refund at epoch 1, which the test forbids.

File: database/plugin/metadata/sqlstore/migrations/pool_deposit_held_test.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+WHERE pool_key_hash = ?`,
+		keyHash,
+	)
+	require.NoError(t, err)
+
+	// Replay only the backfill statement; the ALTER cannot be re-executed on
+	// SQLite and an interrupted upgrade would fail before reaching it anyway.
+	_, err = db.ExecContext(
+		context.Background(),
+		registry[6].SQL["sqlite"].Expand[1],
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply database/plugin/metadata/sqlstore/migrations/pool_deposit_held_test.go]: Valid. Replaced with depositHeldBackfillStatement, which resolves the statement by migration name (pool-registration-deposit-held) and by content (an UPDATE on pool_registration mentioning deposit_held), and requires exactly one match. A statement added to or reordered within v7/sqlite/expand.sql now fails that lookup instead of silently selecting a different statement, and the migration is no longer selected by registry position either.

The finding 1 fix does not add statements to that file: it lives in the migration runner, so v7's expand phase is still the same two statements and registry_test.go's shape assertions are unchanged.

The stale rationale on that test is also gone. It previously said the ALTER could not be re-executed on SQLite, which the runner change makes false; it now states that the backfill is replayed alone so the assertion covers its own IS NULL guard.

File: database/plugin/metadata/store.go [hunk 4]
```diff
@@ -1781,8 +1781,10 @@ type MetadataStore interface {
 	// GetPoolsRetiringAtEpoch returns the pools whose effective retirement
 	// (the latest retirement not cancelled by a later re-registration, as of
 	// the boundary slot) takes effect at the given epoch, along with the
-	// reward account and deposit from their active registration. Used to apply
-	// POOLREAP deposit refunds at the epoch boundary.
+	// reward account from their active registration and the deposit that
+	// registration holds -- the amount its first registration paid, not what
+	// the current protocol parameters would charge. Used to apply POOLREAP
+	// deposit refunds at the epoch boundary.
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply database/plugin/metadata/store.go]: Valid. A registration made after the pool's retirement was reaped is a first registration again and pays the deposit in force at its own slot, which is what `poolRegistrationDepositHeld` implements; the wording overstated it.

Reworded to "the first registration since the pool's most recent completed reap" in `database/plugin/metadata/store.go` and `database/models/pool.go`, and qualified the no-new-deposit rule to a re-registration of a still-registered pool in `ARCHITECTURE.md`, `ledger/poolreap.go`, and the two `DATABASE.md` sites carrying the same phrasing. Documentation only; no behavior change.

File: database/plugin/metadata/sqlstore/migrations/runner.go [hunk 5]
```diff
@@ -609,12 +610,85 @@ func execDDL(
...
+	var found int
+	if err := conn.QueryRowContext(
+		ctx,
+		query,
+		match[1],
+		match[2],
+	).Scan(&found); err != nil {
+		return false
+	}
+	return found != 0
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply database/plugin/metadata/sqlstore/migrations/runner.go]: Half valid, and fixed. `execDDL` now compares the type the live schema reports (`pragma_table_info.type` on SQLite, `information_schema.columns.data_type` on PostgreSQL and MySQL) with the type the statement declares, so a same-named column of another type fails the migration instead of being skipped. That is the line `isMySQLDDLAlreadyAppliedOnConn` already holds for a duplicate index or constraint name. Types are compared without case or length/precision arguments, and through an alias table for the spellings `information_schema` substitutes.

The duplicate-column-error half is rejected. Per-dialect error matching is exactly what the schema query replaces, and it buys nothing: an ADD COLUMN whose column already exists with the declared type cannot succeed on any retry, and a fault that persists still fails the remaining statements, the phase advance, and the contract writes, which run unguarded on the same connection.

Coverage: `TestRunnerReportsAddColumnTypeMismatch` (fails without the type comparison), a declared-type parsing table in `TestAddColumnPatternMatchesShippedMigrations` that also pins every shipped ADD COLUMN to `blob`/`bytea`/`text`, and `TestRunnerReplaysEveryShippedVersionFromExpand` unchanged. Checked against PostgreSQL 16 and MySQL 8: matching type still skips the replay, mismatched type fails with the duplicate-column error. No migration SQL changed, so no checksum changed.

**Review by arepala-uml — CHANGES_REQUESTED** [review]
## Request changes

### Issue: Migration backfills the wrong held deposit for historical re-registrations

`database/plugin/metadata/sqlstore/migrations/v10/sqlite/expand.sql` currently does this:

~~~sql
UPDATE `pool_registration`
SET `deposit_held` = COALESCE(`deposit_amount`, '0')
WHERE `deposit_held` IS NULL;
~~~

This does not reconstruct the deposit the pool actually holds.

Reproduction:

1. A pool initially registers and pays `500000000`.
2. `poolDeposit` changes to `800000000`.
3. The pool re-registers. Cardano does not charge another deposit, so the pool still holds `500000000`.
4. An existing Dingo database runs migration v10.
5. The migration assigns the latest registration `deposit_held = 800000000`.
6. POOLREAP can consequently refund `800000000`, although only `500000000` is held.

This can make an upgraded node disagree with a node synced from genesis and can create or destroy ledger value at the epoch boundary.

#### Regression test

I added this test locally to `database/plugin/metadata/sqlstore/migrations/pool_deposit_held_test.go`:

~~~go
// A migrated database must reconstruct the deposit that the pool still holds,
// rather than copying the protocol parameter recorded on a later
// re-registration. The later registration pays no new pool deposit, so an
// upgraded node must agree with a node that replayed the same history from
// genesis.
func TestDepositHeldBackfillCarriesInitialDepositAcrossReregistration(
	t *testing.T,
) {
	t.Parallel()
	db, runTo := depositHeldBackfillDB(t)
	keyHash := []byte("legacy-pool-key-hash-0000010")
	seedLegacyPoolRegistration(t, db, keyHash, 100, "500000000")

	var poolID int64
	require.NoError(t, db.QueryRowContext(context.Background(), `
SELECT id FROM pool WHERE pool_key_hash = ?`, keyHash).Scan(&poolID))
	_, err := db.ExecContext(context.Background(), `
INSERT INTO pool_registration (
    pool_id, pool_key_hash, added_slot, deposit_amount
) VALUES (?, ?, ?, ?)`,
		poolID, keyHash, 200, "800000000",
	)
	require.NoError(t, err)

	registry, err := migrations.SQLiteRegistry()
	require.NoError(t, err)
	runTo(registry)

	var held sql.NullString
	require.NoError(t, db.QueryRowContext(context.Background(), `
SELECT deposit_held FROM pool_registration
WHERE pool_key_hash = ?
ORDER BY added_slot DESC
LIMIT 1`, keyHash).Scan(&held))
	require.True(t, held.Valid)
	require.Equal(
		t,
		"500000000",
		held.String,
		"re-registration must retain the initial registration's held deposit",
	)
}
~~~

#### Proof

Command:

~~~sh
go test -count=1 \
  -run '^TestDepositHeldBackfillCarriesInitialDepositAcrossReregistration$' \
  ./database/plugin/metadata/sqlstore/migrations
~~~

The test reaches the intended behavioral assertion and fails:

~~~text
expected: "500000000"
actual:   "800000000"

Test: TestDepositHeldBackfillCarriesInitialDepositAcrossReregistration
Messages: re-registration must retain the initial registration's held deposit
FAIL
~~~

This is not a compilation, setup, or fixture failure. It directly demonstrates that migration v10 stores the wrong held deposit.

#### Suggested fix

Replace the blanket SQL copy with a history-aware, resumable migration backfill. Process each pool’s registrations and retirements in chain order using:

~~~text
added_slot, block_index, cert_index
~~~

For each registration:

- Use its own `deposit_amount` when it is the first registration after a completed reap.
- Otherwise carry forward the preceding effective registration’s `deposit_held`.
- Treat a registration before the scheduled retirement epoch as cancellation of the pending retirement.
- Preserve the Mithril synthetic-retirement ordering rules.
- Fail closed with an explicit resync-required error if the available epoch or certificate history cannot determine whether a reap occurred. Do not silently substitute the later registration’s protocol-parameter amount.
- Keep the backfill idempotent and do not overwrite already-populated `deposit_held` values during interrupted-migration replay.

Please include migration tests covering:

- deposit increases before re-registration;
- deposit decreases before re-registration;
- repeated re-registrations;
- pending retirement cancelled by re-registration;
- re-registration after a completed reap;
- rollback/replay or interrupted-backfill behavior;
- incomplete history where the held amount cannot be reconstructed safely.

### Documentation gap: `DATABASE.md` remains inconsistent with the new contract

The PR description says `DATABASE.md` was updated, but it is not part of the current PR diff.

The document still:

- omits `deposit_held` from the `pool_registration` schema row;
- says the refund deposit comes from the latest registration without distinguishing the retained deposit;
- shows `GetPoolsRetiringAtEpoch` selecting `pr.deposit_amount`.

A scoped content check confirms the mismatch:

~~~sh
section=$(sed -n '2364,2425p' DATABASE.md)
if print -r -- "$section" | rg -q 'deposit_held'; then
  print 'PASS: GetPoolsRetiringAtEpoch documentation describes deposit_held'
else
  print 'FAIL: GetPoolsRetiringAtEpoch documentation never mentions deposit_held'
  exit 1
fi
~~~

Result:

~~~text
FAIL: GetPoolsRetiringAtEpoch documentation never mentions deposit_held
~~~

Exit code: `1`.

A permanent unit test that hard-codes documentation prose would be brittle; the appropriate correction is to update `DATABASE.md` and run the existing documentation validation.

#### Suggested documentation fix

Update `DATABASE.md` to describe:

- the new nullable `pool_registration.deposit_held` column;
- the distinction between `deposit_amount` and `deposit_held`;
- the migration and legacy-row behavior;
- `GetPoolsRetiringAtEpoch` using `COALESCE(pr.deposit_held, pr.deposit_amount)`;
- the resync or fail-closed behavior selected for history that cannot be reconstructed safely.

Then run:

~~~sh
make docs-parity
make sql-check
~~~

**Review by arepala-uml — CHANGES_REQUESTED** [review]
## Request changes

### Issue: `ffffe67d` silently regressed the ADD COLUMN replay type-check, and deleted the test that caught it

`51977d1e` ("fix(database): verify column type before skipping replayed ADD COLUMN") added a real guard: an `ALTER TABLE ... ADD COLUMN` failure is only treated as an already-applied replay (needed for restart-safety across an interrupted upgrade) when the *existing* column's type matches the type the statement declares. Without that, a same-named column of another type -- hand-added, or left behind by a foreign schema -- is silently accepted as "already migrated" and the runner proceeds to run the rest of the expand phase against the wrong column.

`ffffe67d` ("test(migrations): keep column helpers with tests") moved the type-comparison helpers (`declaredColumnType`, `normalizeColumnType`, `columnTypeAliases`) out of `runner.go` into `runner_test.go` as orphaned test-only helpers, but never rewired `isSQLiteDDLAlreadyAppliedOnConn` / `isPostgresDDLAlreadyAppliedOnConn` to call them -- both now only check that the column *exists*, not that its type matches. `mysqlColumnAlreadyPresentDefault` (`mysql_ddl_error_default.go`) and the tagged-build duplicate-column branch in `mysql_ddl_error.go` have the same existence-only gap; they were never wired to the type check at all. The regression test that this behavior change should have failed, `TestRunnerReportsAddColumnTypeMismatch`, was deleted in the same commit instead of the code being fixed.

#### Reproduction

1. A table has a pre-existing `note blob` column (hand-added, or left over from a different schema history).
2. A migration's expand phase runs `ALTER TABLE item ADD COLUMN note text`.
3. SQLite raises "duplicate column name: note".
4. `isSQLiteDDLAlreadyAppliedOnConn` confirms only that a column named `note` exists, accepts the failure as an idempotent replay, and the migration continues -- silently -- against a `blob` column instead of the `text` column the migration meant to add.

#### Regression test

I restored this test (deleted in `ffffe67d`) locally to `database/plugin/metadata/sqlstore/migrations/runner_test.go`:

~~~go
// An already-present column is an applied replay only when its type is the one
// the statement declares. A same-named column of another type is a schema the
// migration never produced, so the ALTER must still fail instead of being
// skipped and leaving the rest of the phase to run against the wrong column.
//
// This reproduces the runtime path removed from isSQLiteDDLAlreadyAppliedOnConn
// / isPostgresDDLAlreadyAppliedOnConn / mysqlColumnAlreadyPresentDefault: those
// functions confirm only that the named column exists, not that its type
// matches, so a pre-existing "note blob" column silently satisfies an
// "ADD COLUMN note text" replay and the migration proceeds against the wrong
// column instead of failing.
func TestRunnerReportsAddColumnTypeMismatch(t *testing.T) {
	t.Parallel()
	db := openTestDB(t)
	runner := &Runner{
		DB:      db,
		Dialect: "sqlite",
		Registry: []Migration{
			{
				Version:          1,
				Name:             "create_item",
				BackfillRevision: "1",
				SQL: map[string]SQL{
					"sqlite": {
						Expand: []string{
							"CREATE TABLE IF NOT EXISTS item " +
								"(id INTEGER PRIMARY KEY, note blob)",
						},
					},
				},
			},
			{
				Version:          2,
				Name:             "add_column_other_type",
				BackfillRevision: "1",
				SQL: map[string]SQL{
					"sqlite": {
						Expand: []string{
							"ALTER TABLE `item` ADD COLUMN `note` text",
							"INSERT INTO item (id, note) VALUES (1, 'seeded') " +
								"ON CONFLICT (id) DO NOTHING",
						},
					},
				},
			},
		},
		Locker: NewProcessLocker(),
	}

	err := runner.Run(context.Background())
	require.ErrorContains(t, err, "failed in "+string(PhaseExpand))
	require.ErrorContains(t, err, "statement 1")

	// The statements after the failed ALTER must not have run.
	var rows int
	require.NoError(
		t,
		db.QueryRow("SELECT COUNT(*) FROM item").Scan(&rows),
	)
	require.Zero(t, rows)
}
~~~

#### Proof

Command:

~~~sh
go test -run '^TestRunnerReportsAddColumnTypeMismatch$' \
  ./database/plugin/metadata/sqlstore/migrations/... -v
~~~

Against the current head (`b0e6894a`), the test reaches the intended assertion and fails:

~~~text
=== RUN   TestRunnerReportsAddColumnTypeMismatch
2026/09/03 17:53:02 INFO metadata schema migration complete version=1 name=create_item
2026/09/03 17:53:02 INFO metadata schema migration complete version=2 name=add_column_other_type
    runner_test.go:498: Error Trace: .../runner_test.go:498
        Error:      An error is expected but got nil.
        Test:       TestRunnerReportsAddColumnTypeMismatch
--- FAIL: TestRunnerReportsAddColumnTypeMismatch (0.01s)
FAIL
~~~

The migration reports `version=2 name=add_column_other_type` complete instead of failing in the expand phase -- the type guard is gone at head, not merely undertested.

This is not a compilation, setup, or fixture failure. It directly demonstrates that the replayed `ADD COLUMN` is silently accepted against a column of the wrong type.

#### Suggested fix

Restore the type check in production code (`runner.go`), not the test file, and extend it to the MySQL paths that were apparently never wired to it in the first place:

- Extend `parseAddColumnStatement` to also return the column's declared definition (`table, column, definition, ok`).
- Move `declaredColumnType` / `normalizeColumnType` / `columnTypeAliases` back into `runner.go`, and add a small `addColumnTypeMatches(reported sql.NullString, definition string) bool` helper.
- `isSQLiteDDLAlreadyAppliedOnConn`: query `type` from `pragma_table_info` instead of `SELECT 1`, and compare via `addColumnTypeMatches`.
- `isPostgresDDLAlreadyAppliedOnConn`: query `data_type` from `information_schema.columns` instead of `SELECT 1`, same comparison.
- `mysqlColumnAlreadyPresentDefault` (`mysql_ddl_error_default.go`) and the 1060 branch in `isMySQLDDLAlreadyAppliedOnConn` (`mysql_ddl_error.go`, `dingo_extra_plugins` build): same change -- query `data_type` and compare instead of just checking existence.

I implemented and verified this locally. With the change:

~~~text
=== RUN   TestRunnerReportsAddColumnTypeMismatch
--- PASS: TestRunnerReportsAddColumnTypeMismatch (0.01s)
PASS
~~~

`go build ./...`, `go build -tags dingo_extra_plugins ./...`, `go vet ./database/...`, and the full `./database/plugin/metadata/sqlstore/...` package (including `TestRunnerReplaysEveryShippedVersionFromExpand` for every shipped migration version) all stay green -- this restores prior behavior rather than introducing anything new.

Please also confirm whether `ffffe67d`'s test deletion was intentional (e.g. a rebase conflict resolved by dropping the losing side) so this class of regression -- a failing test deleted instead of fixed -- can be caught before merge going forward.

**Comment by wolf31o2** [comment]
Migration `v10` is also declared on the un-PR'd branch `fix/3313-reward-seed-provenance` as `rewardSeedFailureSchemaRelease`, against this branch's `poolDepositHeldSchemaRelease`. This PR has the only open PR, so it holds v10; the other needs renumbering before it opens.

File: database/plugin/metadata/sqlstore/pool.go [hunk 6]
```diff
@@ -76,6 +76,287 @@ func insertPoolRegistration(
...
+	}
+	// A row written before the deposit_held column existed falls back to its
+	// own deposit_amount, the same rule the v12 backfill applies, so an
+	// interrupted or skipped backfill still reproduces the pre-change refund
+	// instead of collapsing the held amount to zero.
+	source := held
+	if !source.Valid {
+		source = charged
+	}
+	value, err := parseNullUint64("pool registration deposit held", source)
```
Comment (chrisguiney) [inline database/plugin/metadata/sqlstore/pool.go]: **Blocker.** `parseNullUint64` collapses SQL NULL to `0`, so when the previous registration has *both* `deposit_held` and `deposit_amount` NULL, `row.held` becomes a known `0` and `poolRegistrationDepositHeld` returns `&previous.held`. The re-registration then persists `deposit_held = '0'`, and `GetPoolsRetiringAtEpoch`'s `COALESCE(pr.deposit_held, pr.deposit_amount)` reads that authoritative zero instead of falling through to the re-registration's own recorded amount.

The both-NULL state is the documented unknown-deposit case, not a hypothetical: `ledger.calculateCertificateDeposit` returns `nil, nil` for an era with no `CertDepositFunc` and for `ErrIncompatibleProtocolParams`, `SetTransaction` passes that through as an absent `certDeposits` index, and `applyPoolRegistrationCertificate` writes NULL for both columns. `DATABASE.md` states the contract explicitly: "Recording zero for an unknown deposit therefore refunds zero ... (issue #3829). A genuinely recorded zero stays a non-nil zero."

The doc comment on `poolRegistrationDepositHeld` (L149-154) asserts the opposite of what the code does: "A carried-forward amount is always known -- it was read from an earlier row -- so only the branches that return `charged` can be nil." An earlier row's amount is *not* always known.

Reproduction, through the production write path in `database/plugin/metadata/sqlstore`:

```go
// same as writeDepositHeldCert but passes map[int]uint64{} (non-nil, index 0
// absent) so the deposit is unknown rather than zero
writeDepositHeldCertUnknown(t, store, 100, 0, depositHeldRegistration(pool))
writeDepositHeldCert(t, store, 1_100, 0, depositHeldRegistration(pool), 800)
writeDepositHeldCert(t, store, 1_200, 0, depositHeldRetirement(pool, 3), 0)

require.Equal(t, uint64(800), depositHeldRefund(t, store, pool, 3, 3_000))
```

At `909b3806`:

```
first registration deposit_held=sql.NullString{String:"", Valid:false}
re-registration  deposit_held=sql.NullString{String:"0", Valid:true}
Error: Not equal: expected: 0x320  actual: 0x0
```

`main` refunds `800` here (it selected the latest registration's `deposit_amount`), so this is a refund shortfall introduced by the change, not a pre-existing one. It also reproduces on the tree merged with current `origin/main`.

Suggested fix — carry the absence rather than a zero, using the helper the repository already has for exactly this (`parseNullableUint64`: "Use it wherever a stored zero and an unknown value are different answers"):

```go
// poolRegistrationDepositRow.held becomes *uint64
value, err := parseNullableUint64("pool registration deposit held", source)
...
row.held = value
```

and return `previous.held` (already `*uint64`) from the two carry-forward branches instead of `&previous.held`. I applied exactly that in place: the reproduction above passes and every existing `TestPoolDepositHeld*` stays green.

File: database/plugin/metadata/sqlstore/migrations/registry.go [hunk 7]
```diff
@@ -138,11 +141,290 @@ func registryForDialect(dialect string) ([]Migration, error) {
...
+func parsePoolDepositValue(value sql.NullString) (uint64, error) {
+	if !value.Valid || value.String == "" {
+		return 0, nil
+	}
+	return parsePoolDeposit(value.String)
+}
+
+func previousHeld(reg poolDepositRegistration) (uint64, error) {
+	if !reg.held.Valid {
+		return parsePoolDepositValue(reg.amount)
```
Comment (chrisguiney) [inline database/plugin/metadata/sqlstore/migrations/registry.go:436]: **Blocker, same root cause on the migration side.** `parsePoolDepositValue` returns `0, nil` for a NULL `deposit_amount`, so `previousHeld` reports a known zero for a legacy registration whose deposit was never recorded. `backfillPoolDeposits` then writes `strconv.FormatUint(held, 10)` unconditionally (L283), so the *later* re-registration gets `deposit_held = '0'` even though its own `deposit_amount` is a real number.

This is the upgraded-vs-genesis-synced divergence the earlier review flagged, in the direction the new code introduces: before the migration the refund read the latest registration's `deposit_amount`; after it, the column shadows that with `'0'`.

Reproduction in `migrations_test`:

```go
seedLegacyPoolRegistration(t, db, keyHash, 100, nil) // deposit_amount NULL
// second registration at slot 200 with deposit_amount '800000000'
runTo(registry)
```

```
re-registration deposit_held=sql.NullString{String:"0", Valid:true}
```

Suggested fix: make the reconstructed value `*uint64` (nil for an unreconstructable amount) and skip the `UPDATE` when it is nil, leaving `deposit_held` NULL so `COALESCE(pr.deposit_held, pr.deposit_amount)` keeps the pre-change fallback. `TestDepositHeldBackfillCreditsZeroForNullDeposit` pins the current zero only on a *first* registration, where the pre-change refund was also zero, so it does not catch this; please add the re-registration case.

File: database/plugin/metadata/sqlstore/pool_deposit_held_test.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+// amount forward.
+func TestPoolDepositHeldFallsBackToRecordedDeposit(t *testing.T) {
+	t.Parallel()
+	store := newDepositHeldStore(t)
+	depositHeldEpochs(t, store, 5)
+	pool := depositHeldPoolKey(0xa7)
+
+	writeDepositHeldCert(t, store, 100, 0, depositHeldRegistration(pool), 500)
+	_, err := store.writeDB.Exec(`
+UPDATE pool_registration SET deposit_held = NULL WHERE pool_key_hash = ?`,
```
Comment (chrisguiney) [inline database/plugin/metadata/sqlstore/pool_deposit_held_test.go:405]: Non-blocking, but this is the test that was meant to pin the fallback and it does not reach the state the production path actually produces. The `UPDATE` clears `deposit_held` while leaving `deposit_amount = '500'`, so `latestPoolRegistrationBefore`'s `source` falls back to a *valid* `charged` and `parseNullUint64` never sees a NULL. The real unknown-deposit row has both columns NULL — see the blocker on `pool.go`. Setting `deposit_amount = NULL` here as well would have caught it.

File: database/plugin/metadata/sqlstore/migrations/registry.go [hunk 9]
```diff
@@ -138,11 +141,290 @@ func registryForDialect(dialect string) ([]Migration, error) {
...
+			return fmt.Errorf("pool %d registration %d: %w", poolID, reg.id, err)
+		}
+		if havePrevious {
+			retirement, found := latestPoolRetirement(rets, reg.position)
+			if found && poolDepositPositionBeforeOrEqual(previous.position, retirement.position) {
+				epoch, resolved, epochErr := poolDepositEpochAtSlot(ctx, batch, reg.position.slot)
+				if epochErr != nil {
+					return epochErr
+				}
+				if !resolved {
```
Comment (chrisguiney) [inline database/plugin/metadata/sqlstore/migrations/registry.go:265]: Non-blocking consistency note: an unresolvable epoch fails the whole schema upgrade with `resync required` here, while the runtime path in `poolRegistrationDepositHeld` treats the identical condition as fail-open and returns `charged`. Both are defensible, but the two reconstruct the same quantity from the same tables, and the asymmetry is not stated at either site. Worth a sentence on each explaining why an unplaceable slot is fatal during migration and tolerable during live application.

**Review by chrisguiney — DISMISSED** [review]
Reviewed at `909b3806` in an isolated worktree.

Blocking: the carry-forward collapses an *unknown* held deposit into an authoritative zero, in both the runtime write path and the v12 backfill. A pool whose first registration recorded `deposit_amount = NULL` (issue #3829: era with no `CertDepositFunc`, or `ErrIncompatibleProtocolParams`) and which later re-registers is refunded `0` where `main` refunds the re-registration's recorded amount. That is a POOLREAP refund shortfall against `cardano-node` and against a pre-upgrade dingo. Two inline comments with reproductions.

The two non-blocking notes cover the test that was meant to pin this fallback but never reaches the both-NULL state, and the fail-open/fail-closed asymmetry between the runtime and migration epoch lookups.

Both earlier `CHANGES_REQUESTED` items read as resolved at this head: the history-aware backfill is in `registry.go`, and `TestRunnerReportsAddColumnTypeMismatch` is restored in `runner_test.go` with the type check back in `runner.go`. Cubic's MySQL `BOOLEAN`/`tinyint` P1 is addressed by `mysqlColumnTypeMatches`. `DATABASE.md`'s `GetPoolsRetiringAtEpoch` example now selects `COALESCE(pr.deposit_held, pr.deposit_amount)`.

Checks run locally at this head (go1.27.1 / gouroboros v0.202.10):

| check | exit |
| --- | --- |
| `go build ./...` | 0 |
| `go build -tags dingo_extra_plugins ./...` | 0 |
| `go vet ./...` | 0 |
| `gofmt -l .` | 0 new files (2 pre-existing on `main`) |
| `golangci-lint run ./database/... ./ledger/...` | 0 issues |
| `go test ./database/... ./ledger/... -count=1` | 0 |
| merge with current `origin/main` (12 commits ahead), then build + `go test ./database/plugin/metadata/sqlstore/... ./ledger/` | clean, no conflicts; both findings reproduce on the merged tree |

Pipeline at this head: `CodeQL`, `lint`, `govulncheck`, `docker`, `Analyze (actions)`, `Analyze (go)`, `Conventional Commits`, `DCO`, and `go-test` on Linux, Linux-race and macOS are all `success`. `go-test (Windows) (1.26.x)` is `cancelled` (run 34073082590): setup and `go-build` succeeded and the `go-test` step was still `in_progress` after 95 minutes when the run was cancelled. That is a hang or runner timeout, not a test failure, and I cannot attribute it to this change from CI — no log is retained, the three sibling platforms passed at this same head, and the package tests pass locally. It needs a re-run to establish green.

Not run here: Preview Koios parity replay, conformance vectors, devnet, and Antithesis (infrastructure unavailable). The Koios replay is the acceptance check for a deposit-accounting change; the unit tests above do not substitute for it. `nilaway` and `staticcheck` are not installed locally, so `make lint` was replaced by `golangci-lint` directly.

**Comment by wolf31o2** [comment]
Addressed the remaining human-review findings in `25f5a2ba`:

- Preserve unknown pool registration deposits as NULL when both stored values are unavailable, so `COALESCE` falls back to the current registration amount.
- Apply the same NULL-preserving behavior during migration backfill without changing the existing single-row legacy behavior.
- Added runtime and migration regression coverage for unknown deposits.

Validation: `go test ./database/plugin/metadata/sqlstore/... -count=1`.

**Comment by wolf31o2** [comment]
Updated the unknown-deposit migration assertion and added coverage for a later registration with a computable deposit. Focused migration tests pass; please re-review the current head.

**Comment by wolf31o2** [comment]
Fixed the hosted golangci-lint formatting failure in the pool deposit helper. Local golangci-lint and focused migration/SQL-store tests pass; please re-review the current head.

**Review by arepala-uml — APPROVED** [review]
Reviewed at head 25702dbf. My prior CHANGES_REQUESTED finding (migration backfill defaulting an unknown deposit to 0) is fixed. chrisguiney's separate CHANGES_REQUESTED (NULL deposit values collapsing to an authoritative 0, causing a refund shortfall) is also fixed — verified against the exact repro plus a full local `go build`/`go test`, both green. CI green including Windows. Approving.
